### PR TITLE
Migration app / generic app support

### PR DESCRIPTION
**Pull request by valentunn**
Things left for future PRs:
- [ ] Check Generic App
- [x] CheckMetadataHash extension support
- [ ] UI
- [ ] Cloud backups

### CODE REVIEW

File: bindings/metadata_shortener/rust/src/lib.rs [hunk 1]
```diff
@@ -44,6 +51,11 @@ fn Java_io_novafoundation_nova_metadata_1shortener_MetadataShortener_generate_1e
     signed_extras: jbyteArray,
     additional_signed: jbyteArray,
     metadata: jbyteArray,
+    spec_version: jint,
+    spec_name: JString,
+    base58_prefix: jint,
+    decimals: jint,
+    token_symbol: JString,
```
Comment (valentunn) [inline bindings/metadata_shortener/rust/src/lib.rs:58]: I updated extrinsic proof generation to include additional information to match zondax implementation: https://github.com/Zondax/ledger-polkadot-generic-api/blob/master/rust/src/lib.rs#L177

File: feature-ledger-api/src/main/java/io/novafoundation/nova/feature_ledger_api/data/repository/LedgerDerivationPath.kt [hunk 2]
```diff
@@ -9,4 +9,8 @@ object LedgerDerivationPath {
     fun derivationPathSecretKey(chainId: ChainId): String {
         return "$LEDGER_DERIVATION_PATH_KEY.$chainId"
     }
+
+    fun genericDerivationPathSecretKey(): String {
+        return "$LEDGER_DERIVATION_PATH_KEY.Generic"
```
Comment (valentunn) [inline feature-ledger-api/src/main/java/io/novafoundation/nova/feature_ledger_api/data/repository/LedgerDerivationPath.kt:14]: Decided to store DP for generic app accounts the same way we store it for legacy apps. We will need to adapt cloud backups to support this later

File: feature-ledger-impl/src/main/java/io/novafoundation/nova/feature_ledger_impl/presentation/account/addChain/selectLedger/di/AddChainAccountSelectLedgerModule.kt [hunk 3]
```diff
@@ -42,7 +42,7 @@ class AddChainAccountSelectLedgerModule {
     @IntoMap
     @ViewModelKey(AddChainAccountSelectLedgerViewModel::class)
     fun provideViewModel(
-        substrateApplication: SubstrateLedgerApplication,
+        substrateApplication: MigrationSubstrateLedgerApplication,
```
Comment (valentunn) [inline feature-ledger-impl/src/main/java/io/novafoundation/nova/feature_ledger_impl/presentation/account/addChain/selectLedger/di/AddChainAccountSelectLedgerModule.kt:45]: Temproraily untill I integrate relevant app choosing logic based on network state & wallet type (legace / generic ledger wallet)

File: feature-ledger-impl/src/main/java/io/novafoundation/nova/feature_ledger_impl/sdk/application/substrate/newApp/NewSubstrateLedgerApplication.kt [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        val displayVerificationDialog = SubstrateLedgerAppCommon.DisplayVerificationDialog.fromBoolean(confirmAddress)
+
+        val derivationPath = getDerivationPath(chainId, accountIndex)
+        val encodedDerivationPath = encodeDerivationPath(derivationPath)
+
+        val chain = chainRegistry.getChain(chainId)
+        val ss58Prefix = chain.addressPrefix.toShort()
+
+        val payload = encodedDerivationPath + ss58Prefix.littleEndianBytes
+
```
Comment (valentunn) [inline feature-ledger-impl/src/main/java/io/novafoundation/nova/feature_ledger_impl/sdk/application/substrate/newApp/NewSubstrateLedgerApplication.kt:46]: Note application level uses little endian encoding compared to big endian on transport level

File: feature-ledger-impl/src/main/java/io/novafoundation/nova/feature_ledger_impl/sdk/application/substrate/newApp/NewSubstrateLedgerApplication.kt [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        metaId: Long,
+        chainId: ChainId,
+        payload: SignerPayloadExtrinsic
+    ): List<ByteArray> {
+        val payloadBytes = payload.encodedSignaturePayload(hashBigPayloads = false)
+
+        val derivationPath = getDerivationPath(metaId, chainId)
+        val encodedDerivationPath = encodeDerivationPath(derivationPath)
+
+        val encodedTxPayloadLength = payloadBytes.size.toShort().littleEndianBytes
```
Comment (valentunn) [inline feature-ledger-impl/src/main/java/io/novafoundation/nova/feature_ledger_impl/sdk/application/substrate/newApp/NewSubstrateLedgerApplication.kt:99]: Compared to legacy app chunk encoding, we also prepend tx blob length and append metadata proof :
https://github.com/Zondax/ledger-substrate-js/blob/main/src/common.ts#L223

File: feature-ledger-impl/src/main/java/io/novafoundation/nova/feature_ledger_impl/sdk/connection/usb/UsbLedgerConnection.kt [hunk 6]
```diff
@@ -60,29 +63,20 @@ class UsbLedgerConnection(
...
-                Log.w("Ledger", "Successfully sent ${result} bytes to Ledger")
+                Log.w("Ledger", "Successfully sent $result bytes to Ledger")
             }
         }
 
         var somethingRead: Boolean = false
 
         while (true) {
             val responseBuffer = ByteArray(mtu())
-            val result = usbConnection!!.bulkTransfer(endpointIn!!, responseBuffer, responseBuffer.size, 1000)
```
Comment (antonijzelinskij) [inline feature-ledger-impl/src/main/java/io/novafoundation/nova/feature_ledger_impl/sdk/connection/usb/UsbLedgerConnection.kt:85]: Can we provide an argument name here?

**Review by antonijzelinskij — APPROVED** [review]

**Pull request merged**